x-pack/plugins/fleet/server/services/security/uninstall_token_service/index.ts

```diff
@@ -0,0 +1,412 @@
+  SavedObjectsCreatePointInTimeFinderOptions,
+  SavedObjectsBulkUpdateObject,
+  SavedObjectsFindResult,
+} from '@kbn/core-saved-objects-api-server';
+import type {
+  AggregationsMultiBucketAggregateBase,
+  AggregationsTopHitsAggregate,
+} from '@elastic/elasticsearch/lib/api/types';
+import type { EncryptedSavedObjectsClient } from '@kbn/encrypted-saved-objects-plugin/server';
+import type { KibanaRequest } from '@kbn/core-http-server';
+import { SECURITY_EXTENSION_ID } from '@kbn/core-saved-objects-server';
+import { asyncForEach } from '@kbn/std';
+
+import { UNINSTALL_TOKENS_SAVED_OBJECT_TYPE, SO_SEARCH_LIMIT } from '../../../constants';
+import { appContextService } from '../../app_context';
+import { agentPolicyService } from '../../agent_policy';
+
+interface UninstallTokenSOAttributes {
+  policy_id: string;
+  token: string;
+  token_plain: string;
+}
+
+interface UninstallTokenSOAggregationBucket {
+  key: string;
+  latest: AggregationsTopHitsAggregate;
+}
+
+interface UninstallTokenSOAggregation {
+  by_policy_id: AggregationsMultiBucketAggregateBase<UninstallTokenSOAggregationBucket>;
+}
+
+export interface UninstallTokenServiceInterface {
+  getTokenForPolicyId(policyId: string): Promise<string>;
+  getTokensForPolicyIds(policyIds: string[]): Promise<Record<string, string>>;
+  getAllTokens(): Promise<Record<string, string>>;
+  getHashedTokenForPolicyId(policyId: string): Promise<string>;
+  getHashedTokensForPolicyIds(policyIds?: string[]): Promise<Record<string, string>>;
+  getAllHashedTokens(): Promise<Record<string, string>>;
+  generateTokenForPolicyId(policyId: string, force?: boolean): Promise<string>;
+  generateTokensForPolicyIds(policyIds: string[], force?: boolean): Promise<Record<string, string>>;
+  generateTokensForAllPolicies(force?: boolean): Promise<Record<string, string>>;
+  encryptTokens(): Promise<void>;
+}
+
+export class UninstallTokenService implements UninstallTokenServiceInterface {
+  private _soClient: SavedObjectsClientContract | undefined;
+
+  constructor(private esoClient: EncryptedSavedObjectsClient) {}
+
+  /**
+   * gets uninstall token for given policy id
+   *
+   * @param policyId agent policy id
+   * @returns token
+   */
+  public async getTokenForPolicyId(policyId: string): Promise<string> {
+    return (await this.getTokensForPolicyIds([policyId]))[policyId];
+  }
+
+  /**
+   * gets uninstall tokens for given policy ids
+   *
+   * @param policyIds agent policy ids
+   * @returns Record<policyId, token>
+   */
+  public async getTokensForPolicyIds(policyIds: string[]): Promise<Record<string, string>> {
+    let filter = policyIds
+      .map((policyId) => `${UNINSTALL_TOKENS_SAVED_OBJECT_TYPE}.attributes.policy_id: ${policyId}`)
+      .join(' or ');
+    const query: SavedObjectsCreatePointInTimeFinderOptions = {
+      type: UNINSTALL_TOKENS_SAVED_OBJECT_TYPE,
+      perPage: 0,
+      filter,
+      aggs: {
+        by_policy_id: {
+          terms: {
```

Comment:
I am wondering how that aggregation will work for a larger number of policy, do we have to configure the `size` for that aggregation to have all the results?

Reply:
ah, you're right. `terms` only returns 10 by default so would definitely need to add a `size` here. I believe we have users with >1000 policies so maybe `size: 10,000`?

Reply:
yes 10000  will make sense